keps/sig-node/4216-image-pull-per-runtime-class/README.md

```diff
@@ -0,0 +1,611 @@
+cause increased failures in production. See more in the PRR KEP at
+https://git.k8s.io/enhancements/keps/sig-architecture/1194-prod-readiness.
+
+The production readiness review questionnaire must be completed and approved
+for the KEP to move to `implementable` status and be included in the release.
+
+In some cases, the questions below should also have answers in `kep.yaml`. This
+is to enable automation to verify the presence of the review, and to reduce review
+burden and latency.
+
+The KEP must have a approver from the
+[`prod-readiness-approvers`](http://git.k8s.io/enhancements/OWNERS_ALIASES)
+team. Please reach out on the
+[#prod-readiness](https://kubernetes.slack.com/archives/CPNHUMN74) channel if
+you need any help or guidance.
+-->
+
+### Feature Enablement and Rollback
+
+<!--
+This section must be completed when targeting alpha to a release.
+-->
+
+###### How can this feature be enabled / disabled in a live cluster?
+
+<!--
+Pick one of these and delete the rest.
+
+Documentation is available on [feature gate lifecycle] and expectations, as
+well as the [existing list] of feature gates.
+
+[feature gate lifecycle]: https://git.k8s.io/community/contributors/devel/sig-architecture/feature-gates.md
+[existing list]: https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/
+-->
+
+- [x] Feature gate (also fill in values in `kep.yaml`)
+  - Feature gate name: RuntimeClassInCRICalls
+  - Components depending on the feature gate: kubelet
+- [x] Other
+  - Describe the mechanism: `N/A`
+  - Will enabling / disabling the feature require downtime of the control
+    plane? `No`
+  - Will enabling / disabling the feature require downtime or reprovisioning
+    of a node? `No`
+
+###### Does enabling the feature change any default behavior?
+
+No. Aim of this KEP to ensure default behavior is not changed. Any changes to CRI and kubelet will be under feature gate as mentioned in earlier sections.
+
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+
+Yes. All the code changes needed will be under feature gate `RuntimeClassInCriCalls` that will be off by default for alpha.
+Default exisiting behavior of kubelet will be ensured.
+
+###### What happens if we reenable the feature if it was previously rolled back?
+It should be a no-op
+
+###### Are there any tests for feature enablement/disablement?
+
+Yes, this test will be added for alpha.
+
+### Monitoring Requirements
+
+<!--
+This section must be completed when targeting beta to a release.
+
+For GA, this section is required: approvers should be able to confirm the
+previous answers based on experience in the field.
+-->
+
+`Note:` This section will be filled once we move to beta.
+
+###### How can an operator determine if the feature is in use by workloads?
+
+Image pulls for non-default runtime class goes through. Alternately, one could check the status of the feature gate.
```

Comment:
So, the spirit of this question is "I am an operator with 10,000 clusters, and I have this feature enabled. But this feature is causing problems. Who will I break if I turn off this feature?". Thus, the idea is for a sort of global view of when it is in use. We get a little bit of it from the feature gate. But ideally what we have is a metric - say, for a given node how many times a different image was pulled than would have been otherwise (or even better if instrumentation is OK with the volume, would be to be able to identify the specific workloads).

Note this is a beta question, not required for alpha.

Reply:
An easy way to check this is how many images have a non-default runtime handler associated with it. This metric can be added for beta for sure. Updated the doc with the same.